### DISCUSSION TITLE
Safari build

### DISCUSSION THREAD

**Discussion by trvrplk** [AprilSylph/XKit-Rewritten#387] [Tech Office]
Safari recently (last year I believe) switched to the WebExtensions API, and because of this XKit Rewritten runs pretty much flawlessly in it. Following [this official documentation](https://developer.apple.com/documentation/safariservices/safari_web_extensions/converting_a_web_extension_for_safari) I was very easily able to convert the Firefox version to a Safari extension—you just need to have a Mac with Xcode, run a certain command to convert the extension to an Xcode project, [then compile it](https://developer.apple.com/documentation/safariservices/safari_web_extensions/running_your_safari_web_extension). I've had no issues, but there's one problem: because I don't have a developer account, the extension's unsigned, which means I need to re-enable unsigned extensions every time I relaunch Safari. If anyone has a developer account, it'd be really easy and I'm sure helpful to a lot of people to put out an unofficial (or official!) Safari build.

(If anyone wants to run the unsigned version I compiled, [here's a zip with the binary](https://developer.apple.com/documentation/safariservices/safari_web_extensions/running_your_safari_web_extension). Just make sure to enable unsigned extensions or it won't work.)

**Comment by AprilSylph** [comment]
I plan to release to Safari eventually, but as I understand it, Apple would have to review every version I want to publish, and that's a bag of worms I'm not ready to open with how rapidly development is going. It'll be on the table officially when we hit v1.0.0, i.e. every planned feature is built.

**Reply by trvrplk** [reply]
Apple doesn't have to review it unless you release it through the App Store. Of course, iOS requires you to use the App Store, but to release it for MacOS you just need to sign it with a Developer ID when you're packaging it

**Reply by marcustyphoon** [reply]
This method wouldn't include any automatic updates, though, correct? If that's the case, I don't see a huge advantage of doing it; if a user is prepared to keep the extension up to date from GitHub themself, running unsigned isn't as big an additional step.

**Reply by trvrplk** [reply]
No it wouldn't have automatic updates, but the issue with running unsigned is that it can be incredibly tedious. I managed to figure out how to configure the Xcode build so it would run locally for myself without having to reenable it every time I relaunch Safari, but I suspect there are users who'd like to use it and don't know enough about Mac OS development to be able to do that. So updating might not be automatic, but having to rebuild for each update is significantly harder than just removing the old version and installing the new one.

**Reply by AprilSylph** [reply]
I think encouraging non-developers to install via developer methods is not a great idea. I do want to release through the App Store, anyway.

**Reply by trvrplk** [reply]
Yeah that’s a more succinct way of putting what I was trying to say—most people shouldn’t be messing with Xcode

**Reply by marcustyphoon** [reply]
That's the issue, though, right? Even if April distributed a signed binary for people to sideload, it wouldn't have automatic updates, so you would have regular users running old versions, even after the proper Mac app store release, and would have to handle whatever additional support requests and "XKit Rewritten is broken" posts that would result from that. Building the extension with XCode is a pain, definitely, but it does set a bar of "you know what you're getting yourself into and you're not going to expect everything to work perfectly/would be likely to do additional tests in another browser when reporting an issue." I imagine that's the main reason Apple makes you re-enable unsigned extensions in the first place: as a strong reminder that problems you may encounter aren't necessarily problems with the shipping version of the software in question.

**Reply by trvrplk** [reply]
No, the unsigned measures are for security, to ensure that you get a properly authorized binary. Beta software is distributed as signed binaries all the time. (And FWIW I haven’t had any issues running it.)

**Reply by arcanetrivia** [reply]
> Of course, iOS requires you to use the App Store

Does iOS/iPadOS Safari even support extensions?

**Reply by arcanetrivia** [reply]
(ugh, Github, why won't you save the edited comment...) + additional: because I would love to have XKit on my iPad if it does.

**Reply by AprilSylph** [reply]
@lihatt Yep! Safari on iOS 15 supports extensions.

**Reply by arcanetrivia** [reply]
Yay! Good to know.
